### ISSUE TITLE
out of memory activation=softmax loss=categorical_cross_entropy

### ISSUE THREAD

**Issue by vsuthichai** [keras-team/keras#9150]
I've implemented an LSTM encoder decoder architecture to generate paraphrases.  In the decoder output, I'm using a `Dense(vocab_size)` layer with softmax activation and my loss is categorical cross_entropy.  The size of my vocabulary is 70,000+.  The issue is that I'm experiencing out of memory problems training on just a single example on a gpu with 8gb memory.  A one hot sized target sequence is really huge (shape: [batch_size, max_seq_length, vocab_size]) and I'm unable to perform training with even a really small batch size.  Is there another recommended approach I should be using?  Trimming the vocabulary is an option, but I was wondering if there were any tips to modify my network architecture or the loss function.  I guess what I'm asking for.. is there an efficient way for Keras to approximate the softmax instead of doing the softmax over the entire vocabulary.  Something like noise contrastive estimation maybe?  Any suggestions are appreciate!

Here is my model summary:

```
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to
==================================================================================================
input_2 (InputLayer)            (None, 30)           0
__________________________________________________________________________________________________
input_1 (InputLayer)            (None, 30)           0
__________________________________________________________________________________________________
masking_2 (Masking)             (None, 30)           0           input_2[0][0]
__________________________________________________________________________________________________
masking_1 (Masking)             (None, 30)           0           input_1[0][0]
__________________________________________________________________________________________________
embedding_1 (Embedding)         (None, 30, 300)      22399800    masking_1[0][0]
                                                                 masking_2[0][0]
__________________________________________________________________________________________________
bidirectional_1 (Bidirectional) [(None, 1024), (None 3330048     embedding_1[0][0]
__________________________________________________________________________________________________
concatenate_1 (Concatenate)     (None, 1024)         0           bidirectional_1[0][1]
                                                                 bidirectional_1[0][3]
__________________________________________________________________________________________________
concatenate_2 (Concatenate)     (None, 1024)         0           bidirectional_1[0][2]
                                                                 bidirectional_1[0][4]
__________________________________________________________________________________________________
lstm_2 (LSTM)                   [(None, 30, 1024), ( 5427200     embedding_1[1][0]
                                                                 concatenate_1[0][0]
                                                                 concatenate_2[0][0]
__________________________________________________________________________________________________
time_distributed_1 (TimeDistrib (None, 30, 74666)    76532650    lstm_2[0][0]
==================================================================================================
Total params: 107,689,698
Trainable params: 107,689,698
Non-trainable params: 0
```

**Comment by vsuthichai** [comment]
My model code is here:
```
def lstm_keras(args, embeddings, mask_val):
    vocab_size, embedding_size = embeddings.shape

    # Encoder
    encoder_inputs = Input(batch_shape=(None, args.max_encoder_tokens))
    masked_encoder_inputs = Masking(mask_value=mask_val)(encoder_inputs)

    encoder_embeddings = Embedding(vocab_size, embedding_size, weights=[embeddings], trainable=True)
    x = encoder_embeddings(masked_encoder_inputs)

    encoder_outputs, forward_h, forward_c, backward_h, backward_c = Bidirectional(LSTM(args.hidden_size, return_state=True, recurrent_dropout=args.dropout))(x)
    state_h = Concatenate()([forward_h, backward_h])
    state_c = Concatenate()([forward_c, backward_c])
    encoder_states = [state_h, state_c]

    # Decoder
    decoder_inputs = Input(batch_shape=(None, args.max_decoder_tokens))
    masked_decoder_inputs = Masking(mask_value=mask_val)(decoder_inputs)

    #decoder_embeddings = Embedding(vocab_size, embedding_size, weights=[embeddings], trainable=True)
    x = encoder_embeddings(masked_decoder_inputs)

    decoder_lstm, _, _ = LSTM(args.hidden_size * 2, return_sequences=True, return_state=True, recurrent_dropout=args.dropout)(x, initial_state=encoder_states)
    decoder_outputs = TimeDistributed(Dense(vocab_size, activation='softmax'))(decoder_lstm)

    model = Model(inputs=[encoder_inputs, decoder_inputs], outputs=decoder_outputs)
    optimizer = Adam(lr=1e-3, clipnorm=5.)
    model.compile(optimizer=optimizer, loss='categorical_crossentropy')
    return model
```

**Comment by Dref360** [comment]
You have 107 million parameters, scale down your model.

**Comment by rex-yue-wu** [comment]
Basically, the number of your output classes is too big. Maybe you can better introduce your task, but based on my understanding your model does not make sense, where your input is simply a 30-dim vector, but your output is way larger than input, expanding the 3rd dim from 1 to 74666.

**Comment by vsuthichai** [comment]
Apologies, I'm creating a sentence paraphrasing encoder decoder.  Specifically, I have a bidirectional lstm  encoder and another lstm decoder.  I'm using pre-trained embeddings of dimension 300.  Thirty is the maximum sized source sentence length, or maximum time steps.  Each word in each time step is embedded into a 300 dimensional space and fed into a BiLSTM.  The output, state, and reference sentence is then fed into a decoder for training.  The prediction at each step of the decoder is a word from a vocabulary whose size is 74666.  The hidden size of the BiLSTM is also 300, and the hidden size of the decoder LSTM is 600.  I managed to get this working after rewriting it all using the  Tensorflow api with considerable effort (without Keras and despite the large vocabulary size) to produce semi-reasonable paraphrases.  I posted this at an earlier time when I was having the issue in Keras.  Any thoughts on why this wouldn't be working in Keras though?  The network architecture is pretty much the same.

**Comment by rex-yue-wu** [comment]
Correct me if I am wrong
- your input of encoder is a sentence
- each of its word  has been represented as a fixed dimension vector (dim=300) by a pretrained word2vec like embedder. 
- your input in keras is `sentence_in=Input(shape=(None, 30, 300))`. 

Is this input also composed of words in your vocabulary? If so, why not you apply the embedding also for your decoder's target? In this case, your decoder output tensor is of size `batch_size x 30 x 300`, which is the same as your encoder's input.

To decode the actual sentence in your target, you then need a vec2word model that is paired with your pretrained word2vec.

**Comment by vsuthichai** [comment]
Yes you're absolutely correct.  There used to be an embedding on the decoder side.  At the time I wrote it, I decided to get rid of the decoder's embeddings and just share the embedding between the encoder and decoder to reduce the overall parameter size of the model.

I've since moved over to Tensorflow and no longer have shared embeddings.  I do have to two separate embeddings, one for the encoder and the decoder.  And this works fine.

My question was really more towards why this works in Tensorflow, but not Keras due to the memory allocation error?  I would have liked to get it working with Keras to avoid having to dig too deeply into the Tensorflow api.  My network architecture actually has evolved to to a point in Tensorflow where it is different from the way I coded it in Keras.  There are more things like attention, beam search, etc.  However the base implementation of the network remains the same, ie. BiLSTM, encoder with embeddings, decoder with embeddings (embeddings shared with encoder in Keras but not in my Tensorflow implementation but this is not the problem), dense layer, and softmax.  It is essentially the same network in both implementations.
